insonmnia/gatekeeper/gatekeeper.go

```diff
@@ -0,0 +1,351 @@
+				zap.String("value", inTx.Value.String()),
+				zap.String("tx number", inTx.Number.String()),
+				zap.Uint64("block number", inTx.BlockNumber))
+
+			go g.processUnpaidTransaction(ctx, inTx)
+		}
+	}
+
+	g.findScummyTransactions(ctx, inTxs, outTxs)
+	g.logger.Debug("finish transaction processing")
+	return nil
+}
+
+func (g *Gatekeeper) loadTransactions(ctx context.Context) (map[string]*blockchain.GateTx, map[string]*blockchain.GateTx, error) {
+	var err error
+	inTxs, err := g.in.GetPayinTransactions(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	outTxs, err := g.out.GetPayoutTransactions(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	g.logger.Info("loaded transactions",
+		zap.Int("amount of payin transactions", len(inTxs)),
+		zap.Int("amount of payout transactions", len(outTxs)))
+
+	return inTxs, outTxs, nil
+}
+
+// checkDelay verify that tx out of delay
+func (g *Gatekeeper) checkDelay(ctx context.Context, tx *blockchain.GateTx) bool {
+	payinTimestamp, err := g.in.GetGateTransactionTime(ctx, tx)
+	if err != nil {
+		return false
+	}
+	// cast delay, time of payin tx and now time to uint64
+	payinTime := int64(payinTimestamp)
+	delay := int64(g.cfg.Gatekeeper.Delay.Seconds())
+	nowTime := time.Now().UTC().Unix()
+
+	g.logger.Debug("delay check", zap.Int64("delay", payinTime+delay), zap.Int64("nowTime", nowTime))
+	if nowTime < payinTime+delay {
+		return false
+	}
+	return true
+}
+
+// checkUnpaid verify that tx paid already
+func (g *Gatekeeper) isNotPaid(ctx context.Context, tx *blockchain.GateTx) bool {
+	// verify that transaction not payout now
+	txState, err := g.out.GetTransactionState(ctx, tx.From, tx.Value, tx.Number)
+	if err != nil {
+		g.logger.Debug("err while getting tx data", zap.Error(err))
+		return false
+	}
+	return !txState.Paid
+}
+
+func (g *Gatekeeper) isCommitted(ctx context.Context, tx *blockchain.GateTx) bool {
+	// verify that transaction not payout now
+	txState, err := g.out.GetTransactionState(ctx, tx.From, tx.Value, tx.Number)
+	if err != nil {
+		g.logger.Debug("err while getting tx data", zap.Error(err))
+		return false
+	}
+	return txState.CommitTS.Cmp(big.NewInt(0)) != 0
+}
+
+func (g *Gatekeeper) transactionCommittedByMe(ctx context.Context, tx *blockchain.GateTx) bool {
+	// verify that transaction not payout now
+	txState, err := g.out.GetTransactionState(ctx, tx.From, tx.Value, tx.Number)
+	if err != nil {
+		g.logger.Debug("err while getting tx data", zap.Error(err))
+		return false
+	}
+	return txState.Keeper.String() == crypto.PubkeyToAddress(g.key.PublicKey).String()
+}
+
+func (g *Gatekeeper) underLimit(ctx context.Context, tx *blockchain.GateTx) bool {
+	keeper, err := g.out.GetKeeper(ctx, crypto.PubkeyToAddress(g.key.PublicKey))
+	if err != nil {
+		return false
+	}
+
+	spentToday := keeper.SpentToday
+
+	if !keeper.LastDay.IsInt64() {
+		log.Debug("overflowed LastDay value in keeper state")
+		return false
+	}
+	if keeper.LastDay.Int64() < int64(time.Now().Day()) {
```

Comment:
you, again and again, feeling a lot of butt-fire but continue representing time values as big.Int.